src/rmkit/ui/events.cpy

```diff
@@ -0,0 +1,107 @@
+#include <memory>
+#include "../../shared/clockwatch.h"
+#include "../input/events.h"
+#include "../util/signals.h"
+
+namespace ui:
+  PLS_DEFINE_SIGNAL(MOUSE_EVENT, input::SynMotionEvent)
+  class MOUSE_EVENTS:
+    public:
+    MOUSE_EVENT enter
+    MOUSE_EVENT leave
+    MOUSE_EVENT down
+    MOUSE_EVENT up
+    MOUSE_EVENT click
+    MOUSE_EVENT hover
+    MOUSE_EVENT move
+    MOUSE_EVENT long_click
+    MOUSE_EVENT short_click
+    MOUSE_EVENT drag_start
+    MOUSE_EVENT dragging
+    MOUSE_EVENT drag_end
+
+    class MouseGestureManager:
+      public:
+      input::SynMotionEvent prev
+      ClockWatch timer
+      enum STATE { INACTIVE, DOWN, DRAGGING }
+      STATE state = INACTIVE
+
+      // Convert this into a drag event once the x or y delta is > tolerance.
+      // While x and y delta is < tolerance, this is considered a click
+      int drag_start_tolerance = 50
+
+      // Once dragging, only submit events when the x or y delta > tolerance.
+      // This limits the number of dragging events.
+      int dragging_tolerance = 25
+
+      // Delay (in seconds) before click is considered a long_click
+      double long_click_delay_s = 0.5
+
+      void reset():
+        timer = {}
+        state = INACTIVE
+
+      void attach(MOUSE_EVENTS * mouse):
+        mouse->down += PLS_LAMBDA(auto &ev) {
+          reset()
+          if (ev.left):
+            state = DOWN
+            prev = ev
+            timer = {}
+        }
+
+        mouse->up += PLS_LAMBDA(auto &ev) {
+          if state == INACTIVE:
+            return
+          if state == DRAGGING:
+            mouse->drag_end(ev)
+          else if timer.elapsed() >= long_click_delay_s:
+            mouse->long_click(ev)
+          else:
+            mouse->short_click(ev)
+          reset()
+        }
+
+        mouse->move += PLS_LAMBDA(auto &ev) {
+          if state == INACTIVE:
+            return
+          int tolerance = (state == DRAGGING) ? dragging_tolerance : drag_start_tolerance
+          if std::abs(ev.x - prev.x) > tolerance || std::abs(ev.y - prev.y) > tolerance:
+            if state == DRAGGING:
+              mouse->dragging(ev)
+            else:
+              state = DRAGGING
+              mouse->drag_start(prev)
+            prev = ev
+        }
+
+        mouse->leave += PLS_LAMBDA(auto &ev) {
+          if state == DRAGGING:
+            mouse->drag_end(ev)
+          reset()
+        }
+    ;
+
+    unique_ptr<MouseGestureManager> gesture_mgr
+
+    // set up gesture events: drag, long_click, short_click
+    void init_gestures()
+      gesture_mgr = make_unique<MouseGestureManager>()
+      gesture_mgr->attach(this)
```

Comment:
i think this should only be callable once. accidental double call might be weird

Reply:
Agreed. I've been thinking more and I believe there are at least a couple ways to make this automatic

1. A mixin class:

   ```cpp
   template<typename WIDGET>
   class GestureMixin : public WIDGET {
     // combination of MouseGestureManager and events
     class GESTURE_EVENTS {
       MOUSE_EVENT drag_start;
       MOUSE_EVENT dragging;
       MOUSE_EVENT drag_end;
       GESTURE_EVENTS(MOUSE_EVENTS * mouse) {
         attach(mouse);
       }
     };
     // initialize by attaching to existing mouse events
     GESTURE_EVENTS gestures = GESTURE_EVENTS(&this->mouse);
   };

   // Usage -- maybe a little ugly, but very clear IMO
   auto button = ui::GestureMixin<ui::Button>(0, 0, 200, 200, "Hello")
   ```

2. A `mouse.down` listener:

   ```cpp
   // in the widget constructor
   mouse.down += PLS_LAMBDA(auto & ev) {
     // this part is pretty bad, but the user wouldn't have to know
     if !gesture_mgr && (!mouse.drag_start.cbs.empty() || !mouse.dragging.empty() ...):
       gesture_mgr = make_unique<GestureManager>()
       gesture_mgr->attach(mouse)
   }
   ```

Reply:
the mixin is nicer, but requires the dev to turn it on when they want to use drag gestures, so prefer method that the dev doesn't need to know anything about (just transparently works)